core/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java

```diff
@@ -118,53 +129,55 @@ void executeSearch(Queue<SearchRequestSlot> requests, AtomicArray<MultiSearchRes
+         * we recurse.
          */
-        searchAction.execute(request.request, new ActionListener<SearchResponse>() {
-            @Override
-            public void onResponse(final SearchResponse searchResponse) {
-                handleResponse(request.responseSlot, new MultiSearchResponse.Item(searchResponse, null));
-                executeSearchLoop();
-            }
-
-            @Override
-            public void onFailure(final Exception e) {
-                handleResponse(request.responseSlot, new MultiSearchResponse.Item(null, e));
-                executeSearchLoop();
+        final AtomicReference<SearchRequestSlot> next = new AtomicReference<>();
+        final Thread thread = Thread.currentThread();
+        do {
+            final SearchRequestSlot current;
+            if (next.get() != null) {
+                current = next.get();
+                next.set(null);
+            } else {
+                current = request;
             }
-
-            private void handleResponse(final int responseSlot, final MultiSearchResponse.Item item) {
-                responses.set(responseSlot, item);
-                if (responseCounter.decrementAndGet() == 0) {
-                    assert requests.isEmpty();
-                    finish();
+            searchAction.execute(current.request, new ActionListener<SearchResponse>() {
+                @Override
+                public void onResponse(final SearchResponse searchResponse) {
+                    handleResponse(current.responseSlot, new MultiSearchResponse.Item(searchResponse, null));
                 }
-            }
 
-            private void finish() {
-                listener.onResponse(new MultiSearchResponse(responses.toArray(new MultiSearchResponse.Item[responses.length()])));
-            }
+                @Override
+                public void onFailure(final Exception e) {
+                    handleResponse(current.responseSlot, new MultiSearchResponse.Item(null, e));
+                }
 
-            private void executeSearchLoop() {
-                SearchRequestSlot next;
-                while ((next = requests.poll()) != null) {
-                    final int nextResponseSlot = next.responseSlot;
-                    searchAction.execute(next.request, new ActionListener<SearchResponse>() {
-                        @Override
-                        public void onResponse(SearchResponse searchResponse) {
-                            handleResponse(nextResponseSlot, new MultiSearchResponse.Item(searchResponse, null));
+                private void handleResponse(final int responseSlot, final MultiSearchResponse.Item item) {
+                    responses.set(responseSlot, item);
+                    if (responseCounter.decrementAndGet() == 0) {
+                        assert requests.isEmpty();
+                        finish();
+                    } else {
+                        if (thread == Thread.currentThread()) {
```

Comment:
Can we instead always do: `next.set(requests.poll());`? I think this would make the code simpler?